ruby/ext/google/protobuf_c/storage.c

```diff
@@ -447,6 +428,22 @@ MessageLayout* create_layout(const upb_msgdef* msgdef) {
 
   layout->fields = ALLOC_N(MessageField, nfields);
 
+  size_t hasbit = sizeof(void*) * 8;
```

Comment:
@zanker do you remember the reason for setting this to `sizeof(void*) * 8` and not 0?

Reply:
I thought it had to do with pointer alignment, but apparently past me was smart and wrote a helpful comment. From `encode_decode.c`

```
// The offset we pass to UPB points to the start of the Message, rather than the start of where our data is stored.
// Which works out to the size of a pointer. Since we can't pass arbitrary pointers to UPB, we bump it by the size*8
// which lets us reuse the UPB writers for primitives.
int32_t hasbit = hasbit_off == MESSAGE_FIELD_NO_HASBIT ? -1 : hasbit_off + sizeof(void*) * 8;
```
The general gist of the presence checking, is that it's a bit hackish because I'm trying to reuse as much of the existing UPB code as possible. There's definitely nicer ways to do it, but you end up reimplementing more code.

Reply:
Well, starting at `sizeof(void*) * 8` sets the first hasbit at offset 64, which will allocate the 9th byte in the data part of `MessageHeader`. So effectively, there is no difference whether you start from 0, or 64 or any other value.

As for the code
```
int32_t hasbit = hasbit_off == MESSAGE_FIELD_NO_HASBIT ? -1 : hasbit_off + sizeof(void*) * 8;
```
it should be `sizeof(MessageHeader)` instead of `sizeof(void*)` as you want the offset to the data part of the message type and skip the number of bytes needed for `MessageHeader`

Reply:
If you're able to get it to work by simplifying it, then I would definitely do so. You should also add some tests that encode and decode a message then check the `has_foo?` fields to verify. The main thing, is that it's all setup to align to the offsets used by: https://github.com/zanker/protobuf-1/blob/af38326c9a7ceab44b148a80ed8e1e734821733f/ruby/ext/google/protobuf_c/upb.c#L3282, not by `MessageLayout`.

Reply:
Ahh.. I see. Looks like that has changed in the recent version of UPB. https://github.com/google/protobuf/blob/master/ruby/ext/google/protobuf_c/upb.c#L4311

Reply:
Ah! Awesome. That should make life easier then.

It looks like the version of UPB got upgraded with the unknown field change, and so you now have things like `upb_sethasbit` and `upb_readhasbit`. That should make it a lot easier and let you clean up the code. A lot of the issues I had were that I had to reimplement them to match what UPB did.

Reply:
Don't know if I can use `upb_sethasbit` or `upb_readhasbit` without switching to using UPB for message layout and the in-memory representation of a message. And that to me sounds like more work that this PR is intended for.

Between, @haberman is the eventual plan to use UPB for message layout and the in-memory representation?